module/modmanager/manager_test.go

```diff
@@ -459,6 +460,59 @@ func TestModuleReloading(t *testing.T) {
 		// Assert that RemoveOrphanedResources was called once.
 		test.That(t, dummyRemoveOrphanedResourcesCallCount.Load(), test.ShouldEqual, 1)
 	})
+	t.Run("immediate crash is not restarted", func(t *testing.T) {
```

Comment:
Oooh... good catch. I totally overlooked that, and suspect Reconfigure gets the "cannot reconfigure.... does not exist" error. Probably need to mark the module as "failed" in the module struct itself, rather than deleting from the list.

Reply:
Ah good suggestion! I added a test where:
1.  a robot is created with a regular module with one resource
2.  the module is reconfigured to a crashing module
3.  the resource is lost
4.  the module is reconfigured to the old working module
5.  the resource is regained

> and suspect Reconfigure gets the "cannot reconfigure.... does not exist" error

There was no error like this, and the test pretty much just worked. But, importantly, step (4) in this case is actually "adding" a module, not "reconfiguring" the original one, as the original one was deleted in (2)'s crash. It seems like `h` can be added to this "new" module, so everything works out. I could probably refactor this PR to put modules in a "failed" state when they crash, but I think that will be less trivial.

Reply:
So between your "cleanup" doing a `delete(mgr.modules, mod.name)` and the very top of mgr.Reconfigure() having a check for the module existing in that mgr.modules map, I couldn't figure out how in the world it was NOT failing there, as obviously the map would no longer contain the module.

I finally found it now, because I wasn't aware of https://github.com/viamrobotics/rdk/pull/2437 which completely changes how Reconfigure was working before that. So the robot's config is not being diffed against the last actual version from the cloud, but against a live/generated virtual config... and the modules part of that is made FROM mgr.modules, so deleting it there (in quick-crash cleanup) effectively ALSO deletes it from the config (this is the key part I didn't know about), and thus the next reconfigure is always an Add() not a Reconfigure().

It makes sense now that I'm aware of that, but boy did I feel like I was going crazy for a short bit here, as it seemed impossible (and I believe would've been) under the old config-diff paradigm.

Reply:
Ah excellent find; thanks for looking into that! Yeah things definitely changed with #2437. But, the current behavior makes more intuitive sense to me (module was "removed" in crash, and can therefore only be "added" as a separate module now).